{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fcf77a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "import subprocess\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d113ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReadAndFilter(file, files_dir, chrom2use):\n",
    "    sample = pd.read_csv(files_dir+file, sep=\"\\t\")\n",
    "    sample[\"VC\"] = sample[\"chrom\"].astype(str)+\"_\"+sample[\"position\"].astype(str)+\"_\"+sample[\"ref\"].astype(str)+\"_\"+sample[\"var\"].astype(str)\n",
    "    sample[\"T_DP\"] = sample[\"tumor_reads1\"]+ sample[\"tumor_reads2\"]\n",
    "    sample[\"N_DP\"] = sample[\"normal_reads1\"]+sample[\"normal_reads2\"]\n",
    "    sample[\"N_VAF\"] = sample[\"normal_var_freq\"].str.split(\"%\", expand=True)[0].astype(float)/100\n",
    "    sample[\"N_ALT\"] = sample[\"N_VAF\"]*sample[\"N_DP\"]\n",
    "    sample[\"N_ALT\"] = sample[\"N_ALT\"].astype(int)\n",
    "    sample[\"T_VAF\"] = sample[\"tumor_var_freq\"].str.split(\"%\", expand=True)[0].astype(float)/100\n",
    "    sample[\"T_ALT\"] = sample[\"T_VAF\"]*sample[\"T_DP\"]\n",
    "    sample[\"T_ALT\"] = sample[\"T_ALT\"].astype(int)\n",
    "    sample[\"chrom\"] = sample[\"chrom\"].astype(str)\n",
    "    sample = sample[sample[\"chrom\"].isin(chrom2use)]\n",
    "    col2show = [\"VC\", \"N_DP\", \"N_VAF\", \"N_ALT\", \"T_DP\", \"T_VAF\",\"T_ALT\",  \"somatic_status\", \"somatic_p_value\"]\n",
    "    return sample[col2show][(sample[\"N_VAF\"]==0)&(sample[\"somatic_p_value\"]<0.05)#&(sample[\"T_VAF\"]>=0.1)\n",
    "                            #&(sample[\"N_DP\"]>50)&(sample[\"T_DP\"]>50)&(sample[\"T_ALT\"]>5)\n",
    "                            &(sample[\"somatic_status\"]==\"Somatic\")].sort_values(by=\"somatic_p_value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4b49c915",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom2use = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15',\n",
    "             '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943c7e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-define FASTQ file directory (i.e., filed_dir)\n",
    "def SNP(sample_name, chrom2use, files_dir = \"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/\"):\n",
    "    snp = ReadAndFilter(sample_name, files_dir = files_dir+\"filtered_vcfs/\", chrom2use=chrom2use)\n",
    "    return snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b04c4c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-define FASTQ file directory (i.e., filed_dir)\n",
    "def SNPandIndel(sample_name, chrom2use, files_dir = \"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/\"):\n",
    "    snp = ReadAndFilter(sample_name, files_dir = files_dir+\"filtered_vcfs/\", chrom2use=chrom2use)\n",
    "    snp[\"type\"] = \"snp\"\n",
    "    indel = ReadAndFilter(sample_name+\".indel\", files_dir = files_dir+\"vcfs/\", chrom2use=chrom2use)\n",
    "    indel[\"type\"] = \"indel\"\n",
    "    combined = pd.concat([snp, indel])\n",
    "    return combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6eace42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GetCoverageAtPos(row, bam_file):\n",
    "    command = [\"samtools\", \"depth\", bam_file, \"-r\", f\"{row.chrom}:{row.pos}-{row.pos}\"]\n",
    "    result = subprocess.run(command, stdout=subprocess.PIPE, text=True)\n",
    "    if result.returncode == 0:\n",
    "        lines = result.stdout.strip().split('\\n')\n",
    "        if lines:\n",
    "            coverage = int(lines[0].split('\\t')[2])\n",
    "            return coverage\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa1c1cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Position2gene(row, gene2regions):\n",
    "    df = gene2regions[(gene2regions[\"chrom\"]==row.chrom)\n",
    "                      &(gene2regions[\"start\"]<=int(row.pos))\n",
    "                      &(gene2regions[\"end\"]>=int(row.pos))]\n",
    "    genes = list(df[\"gene\"].values)\n",
    "    if len(genes)>0:\n",
    "        return genes[0]\n",
    "    else:\n",
    "        return \"nan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f782d8eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BAM file directory\n",
    "bam_folder = \"/home/bioit/mlarmuse/Documents/Data/Almac_project/CopenHagen_oligo/tNGS_bams/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5f2553e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient ID4\n",
    "pat_id = \"ID4\"\n",
    "num_id = \"4\"\n",
    "ID4ids = [\"HR41\", \"HR42\", \"HR43\", \"HR44\", \"HR45\", \"HR4MLN1\", \"HR4MLN2\", \"HR4MLN3\"]\n",
    "ID4_variants = pd.DataFrame()\n",
    "for f in ID4ids:\n",
    "    var = SNPandIndel(f+\".somaticFilter\", chrom2use=chrom2use)\n",
    "    var[\"sampleID\"] = f\n",
    "    ID4_variants = pd.concat([ID4_variants, var], axis=0)\n",
    "    \n",
    "ID4_df = ID4_variants.groupby(['VC'])['sampleID'].agg(list).reset_index()\n",
    "ID4_df[\"multiple\"] = ID4_df[\"sampleID\"].str.len()\n",
    "ID4_df[\"MLN\"]  = [1 if any(\"MLN\" in y for y in x) else 0 for x in ID4_df[\"sampleID\"]]\n",
    "ID4_df = ID4_df[(ID4_df[\"multiple\"]>1)&(ID4_df[\"MLN\"]==1)]\n",
    "ID4_df[\"chrom\"] = ID4_df[\"VC\"].str.split(\"_\", expand=True)[0]\n",
    "ID4_df[\"pos\"] = ID4_df[\"VC\"].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "ID4_T_bam = [\"OT-P-\"+pat_id+\"-T-\"+x+\"-KH20191030-C320191031.clip.overlapped.bam\" for x in ID4ids]\n",
    "ID4_N_bam = [\"OT-P-\"+pat_id+\"-N-\"+\"HR\"+num_id+\"NT\"+\"-KH20191030-C320191031.clip.overlapped.bam\"]\n",
    "ID4_T_N_bam = ID4_T_bam+ID4_N_bam\n",
    "\n",
    "for f in ID4_T_N_bam:\n",
    "    name = f.split(\"-\")[4]\n",
    "    ID4_df[name] = ID4_df.apply(GetCoverageAtPos, bam_file=bam_folder+pat_id+\"/\"+f, axis=1)\n",
    "\n",
    "col2use = [c for c in ID4_df.columns if \"HR4\" in c]\n",
    "ID4_df = ID4_df[ID4_df[col2use].min(axis=1)>=50]\n",
    "\n",
    "ID4_res = pd.merge(ID4_df, ID4_variants, on=[\"VC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc18fb2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR41\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"])),\n",
    "      len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR44\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR45\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"])))\n",
    "\n",
    "print(len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR41\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"])),\n",
    "      len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR44\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR45\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"])))\n",
    "\n",
    "print(len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR41\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"])),\n",
    "      len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR44\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"])),\n",
    "     len(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR45\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56457d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"]),\n",
    "     set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN1\"]))\n",
    "\n",
    "print(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"]),\n",
    "     set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN2\"]))\n",
    "\n",
    "print(set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR42\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"]),\n",
    "     set(ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR43\"]).intersection(\n",
    "    ID4_res[\"VC\"][ID4_res[\"sampleID_y\"]==\"HR4MLN3\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf137e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient ID5\n",
    "pat_id = \"ID5\"\n",
    "num_id = \"5\"\n",
    "ID5ids = [\"HR51\", \"HR52\", \"HR53\", \"HR54\", \"HR5MLN1\", \"HR5MLN2\", \"HR5MLN3\"]\n",
    "ID5_variants = pd.DataFrame()\n",
    "for f in ID5ids:\n",
    "    var = SNPandIndel(f+\".somaticFilter\", chrom2use=chrom2use)\n",
    "    var[\"sampleID\"] = f\n",
    "    ID5_variants = pd.concat([ID5_variants, var], axis=0)\n",
    "    \n",
    "ID5_df = ID5_variants.groupby(['VC'])['sampleID'].agg(list).reset_index()\n",
    "ID5_df[\"multiple\"] = ID5_df[\"sampleID\"].str.len()\n",
    "ID5_df[\"MLN\"]  = [1 if any(\"MLN\" in y for y in x) else 0 for x in ID5_df[\"sampleID\"]]\n",
    "ID5_df = ID5_df[(ID5_df[\"multiple\"]>1)&(ID5_df[\"MLN\"]==1)]\n",
    "ID5_df[\"chrom\"] = ID5_df[\"VC\"].str.split(\"_\", expand=True)[0]\n",
    "ID5_df[\"pos\"] = ID5_df[\"VC\"].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "ID5_T_bam = [\"OT-P-\"+pat_id+\"-T-\"+x+\"-KH20191030-C320191031.clip.overlapped.bam\" for x in ID5ids]\n",
    "ID5_N_bam = [\"OT-P-\"+pat_id+\"-N-\"+\"HR\"+num_id+\"NT\"+\"-KH20191030-C320191031.clip.overlapped.bam\"]\n",
    "ID5_T_N_bam = ID5_T_bam+ID5_N_bam\n",
    "\n",
    "for f in ID5_T_N_bam:\n",
    "    name = f.split(\"-\")[4]\n",
    "    ID5_df[name] = ID5_df.apply(GetCoverageAtPos, bam_file=bam_folder+pat_id+\"/\"+f, axis=1)\n",
    "\n",
    "col2use = [c for c in ID5_df.columns if \"HR5\" in c]\n",
    "ID5_df = ID5_df[ID5_df[col2use].min(axis=1)>=50]\n",
    "\n",
    "ID5_res = pd.merge(ID5_df, ID5_variants, on=[\"VC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f977b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR51\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN1\"])),\n",
    "      len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR52\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN1\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR53\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN1\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR54\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN1\"])))\n",
    "\n",
    "print(len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR51\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN2\"])),\n",
    "      len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR52\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN2\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR53\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN2\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR54\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN2\"])))\n",
    "\n",
    "print(len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR51\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN3\"])),\n",
    "      len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR52\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN3\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR53\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN3\"])),\n",
    "     len(set(ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR54\"]).intersection(\n",
    "    ID5_res[\"VC\"][ID5_res[\"sampleID_y\"]==\"HR5MLN3\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7fef2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient ID6\n",
    "pat_id = \"ID6\"\n",
    "num_id = \"6\"\n",
    "ID6ids = [\"HR61\", \"HR62\", \"HR63\", \"HR64\", \"HR6MLN1\", \"HR6MLN2\", \"HR6MLN3\"]\n",
    "ID6_variants = pd.DataFrame()\n",
    "for f in ID6ids:\n",
    "    var = SNPandIndel(f+\".somaticFilter\", chrom2use=chrom2use)\n",
    "    var[\"sampleID\"] = f\n",
    "    ID6_variants = pd.concat([ID6_variants, var], axis=0)\n",
    "    \n",
    "ID6_df = ID6_variants.groupby(['VC'])['sampleID'].agg(list).reset_index()\n",
    "ID6_df[\"multiple\"] = ID6_df[\"sampleID\"].str.len()\n",
    "ID6_df[\"MLN\"]  = [1 if any(\"MLN\" in y for y in x) else 0 for x in ID6_df[\"sampleID\"]]\n",
    "ID6_df = ID6_df[(ID6_df[\"multiple\"]>1)&(ID6_df[\"MLN\"]==1)]\n",
    "ID6_df[\"chrom\"] = ID6_df[\"VC\"].str.split(\"_\", expand=True)[0]\n",
    "ID6_df[\"pos\"] = ID6_df[\"VC\"].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "ID6_T_bam = [\"OT-P-\"+pat_id+\"-T-\"+x+\"-KH20191030-C320191031.clip.overlapped.bam\" for x in ID6ids]\n",
    "ID6_N_bam = [\"OT-P-\"+pat_id+\"-N-\"+\"HR\"+num_id+\"NT\"+\"-KH20191030-C320191031.clip.overlapped.bam\"]\n",
    "ID6_T_N_bam = ID6_T_bam+ID6_N_bam\n",
    "\n",
    "for f in ID6_T_N_bam:\n",
    "    name = f.split(\"-\")[4]\n",
    "    ID6_df[name] = ID6_df.apply(GetCoverageAtPos, bam_file=bam_folder+pat_id+\"/\"+f, axis=1)\n",
    "\n",
    "col2use = [c for c in ID6_df.columns if \"HR6\" in c]\n",
    "ID6_df = ID6_df[ID6_df[col2use].min(axis=1)>=50]\n",
    "\n",
    "ID6_res = pd.merge(ID6_df, ID6_variants, on=[\"VC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95459b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR61\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN1\"])),\n",
    "      len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR62\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN1\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR63\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN1\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR64\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN1\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR65\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN1\"])))\n",
    "\n",
    "print(len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR61\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN2\"])),\n",
    "      len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR62\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN2\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR63\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN2\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR64\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN2\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR65\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN2\"])))\n",
    "\n",
    "print(len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR61\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN3\"])),\n",
    "      len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR62\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN3\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR63\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN3\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR64\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN3\"])),\n",
    "     len(set(ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR65\"]).intersection(\n",
    "    ID6_res[\"VC\"][ID6_res[\"sampleID_y\"]==\"HR6MLN3\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9680f8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient ID9\n",
    "pat_id = \"ID9\"\n",
    "num_id = \"9\"\n",
    "ID9ids = [\"HR91\", \"HR92\", \"HR93\", \"HR94\", \"HR95\", \"HR9MLN1\", \"HR9MLN2\"]\n",
    "ID9_variants = pd.DataFrame()\n",
    "for f in ID9ids:\n",
    "    var = SNPandIndel(f+\".somaticFilter\", chrom2use=chrom2use)\n",
    "    var[\"sampleID\"] = f\n",
    "    ID9_variants = pd.concat([ID9_variants, var], axis=0)\n",
    "    \n",
    "ID9_df = ID9_variants.groupby(['VC'])['sampleID'].agg(list).reset_index()\n",
    "ID9_df[\"multiple\"] = ID9_df[\"sampleID\"].str.len()\n",
    "ID9_df[\"MLN\"]  = [1 if any(\"MLN\" in y for y in x) else 0 for x in ID9_df[\"sampleID\"]]\n",
    "ID9_df = ID9_df[(ID9_df[\"multiple\"]>1)&(ID9_df[\"MLN\"]==1)]\n",
    "ID9_df[\"chrom\"] = ID9_df[\"VC\"].str.split(\"_\", expand=True)[0]\n",
    "ID9_df[\"pos\"] = ID9_df[\"VC\"].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "ID9_T_bam = [\"OT-P-\"+pat_id+\"-T-\"+x+\"-KH20191030-C320191031.clip.overlapped.bam\" for x in ID9ids]\n",
    "ID9_N_bam = [\"OT-P-\"+pat_id+\"-N-\"+\"HR\"+num_id+\"NT\"+\"-KH20191030-C320191031.clip.overlapped.bam\"]\n",
    "ID9_T_N_bam = ID9_T_bam+ID9_N_bam\n",
    "\n",
    "for f in ID9_T_N_bam:\n",
    "    name = f.split(\"-\")[4]\n",
    "    ID9_df[name] = ID9_df.apply(GetCoverageAtPos, bam_file=bam_folder+pat_id+\"/\"+f, axis=1)\n",
    "    \n",
    "col2use = [c for c in ID9_df.columns if \"HR9\" in c]\n",
    "ID9_df = ID9_df[ID9_df[col2use].min(axis=1)>=50]\n",
    "\n",
    "ID9_res = pd.merge(ID9_df, ID9_variants, on=[\"VC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067790ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR91\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN1\"])),\n",
    "      len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR92\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN1\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR93\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN1\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR94\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN1\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR95\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN1\"])))\n",
    "\n",
    "print(len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR91\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN2\"])),\n",
    "      len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR92\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN2\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR93\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN2\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR94\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN2\"])),\n",
    "     len(set(ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR95\"]).intersection(\n",
    "    ID9_res[\"VC\"][ID9_res[\"sampleID_y\"]==\"HR9MLN2\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63226c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient ID10\n",
    "pat_id = \"ID10\"\n",
    "num_id = \"10\"\n",
    "ID10ids = [\"HR101\", \"HR102\", \"HR103\", \"HR104\", \"HR105\", \"HR106\", \"HR10MLN1\", \"HR10MLN2\"]\n",
    "ID10_variants = pd.DataFrame()\n",
    "for f in ID10ids:\n",
    "    var = SNPandIndel(f+\".somaticFilter\", chrom2use=chrom2use)\n",
    "    var[\"sampleID\"] = f\n",
    "    ID10_variants = pd.concat([ID10_variants, var], axis=0)\n",
    "    \n",
    "ID10_df = ID10_variants.groupby(['VC'])['sampleID'].agg(list).reset_index()\n",
    "ID10_df[\"multiple\"] = ID10_df[\"sampleID\"].str.len()\n",
    "ID10_df[\"MLN\"]  = [1 if any(\"MLN\" in y for y in x) else 0 for x in ID10_df[\"sampleID\"]]\n",
    "ID10_df = ID10_df[(ID10_df[\"multiple\"]>1)&(ID10_df[\"MLN\"]==1)]\n",
    "ID10_df[\"chrom\"] = ID10_df[\"VC\"].str.split(\"_\", expand=True)[0]\n",
    "ID10_df[\"pos\"] = ID10_df[\"VC\"].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "ID10_T_bam = [\"OT-P-\"+pat_id+\"-T-\"+x+\"-KH20191030-C320191031.clip.overlapped.bam\" for x in ID10ids]\n",
    "ID10_N_bam = [\"OT-P-\"+pat_id+\"-N-\"+\"HR\"+num_id+\"NT\"+\"-KH20191030-C320191031.clip.overlapped.bam\"]\n",
    "ID10_T_N_bam = ID10_T_bam+ID10_N_bam\n",
    "\n",
    "for f in ID10_T_N_bam:\n",
    "    name = f.split(\"-\")[4]\n",
    "    ID10_df[name] = ID10_df.apply(GetCoverageAtPos, bam_file=bam_folder+pat_id+\"/\"+f, axis=1)\n",
    "\n",
    "col2use = [c for c in ID10_df.columns if \"HR10\" in c]\n",
    "ID10_df = ID10_df[ID10_df[col2use].min(axis=1)>=50]\n",
    "\n",
    "ID10_res = pd.merge(ID10_df, ID10_variants, on=[\"VC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943bf541",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR101\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])),\n",
    "      len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR102\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR103\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR104\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR105\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])),\n",
    "      len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR106\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN1\"])))\n",
    "\n",
    "print(len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR101\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])),\n",
    "      len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR102\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR103\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR104\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR105\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])),\n",
    "     len(set(ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR106\"]).intersection(\n",
    "    ID10_res[\"VC\"][ID10_res[\"sampleID_y\"]==\"HR10MLN2\"])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de0c35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene to regions (hg19) - coding regions\n",
    "hgtable = pd.read_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/gene_table.gz\", sep=\"\\t\")\n",
    "hgtable = hgtable[['name2', \"chrom\", \"txStart\", \"txEnd\"]].drop_duplicates()\n",
    "hgtable.columns = [\"gene\", \"chrom\", \"start\", \"end\"]\n",
    "hgtable[\"chrom\"] = hgtable[\"chrom\"].str[3:]\n",
    "hgtable[\"size\"] = hgtable[\"end\"]-hgtable[\"start\"]\n",
    "hgtable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d4412e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID4_res[\"genes\"] = ID4_res.apply(Position2gene, gene2regions=hgtable, axis=1)\n",
    "ID5_res[\"genes\"] = ID5_res.apply(Position2gene, gene2regions=hgtable, axis=1)\n",
    "ID6_res[\"genes\"] = ID6_res.apply(Position2gene, gene2regions=hgtable, axis=1)\n",
    "ID9_res[\"genes\"] = ID9_res.apply(Position2gene, gene2regions=hgtable, axis=1)\n",
    "ID10_res[\"genes\"] = ID10_res.apply(Position2gene, gene2regions=hgtable, axis=1)\n",
    "\n",
    "ID4_res = ID4_res.drop('sampleID_x', axis=1)\n",
    "ID5_res = ID5_res.drop('sampleID_x', axis=1)\n",
    "ID6_res = ID6_res.drop('sampleID_x', axis=1)\n",
    "ID9_res = ID9_res.drop('sampleID_x', axis=1)\n",
    "ID10_res = ID10_res.drop('sampleID_x', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f9fdbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2use = []\n",
    "for var in set(ID4_res[\"VC\"]):\n",
    "    if max(ID4_res[ID4_res[\"VC\"]==var][\"T_VAF\"])>0.05:\n",
    "           var2use.append(var)\n",
    "ID4_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]][\n",
    "    (ID4_res[\"VC\"].isin(var2use))\n",
    "].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8077e4fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2use = []\n",
    "for var in set(ID5_res[\"VC\"]):\n",
    "    if max(ID5_res[ID5_res[\"VC\"]==var][\"T_VAF\"])>0.05:\n",
    "           var2use.append(var)\n",
    "ID5_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]][\n",
    "    (ID5_res[\"VC\"].isin(var2use))\n",
    "].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "697f3ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2use = []\n",
    "for var in set(ID6_res[\"VC\"]):\n",
    "    if max(ID6_res[ID6_res[\"VC\"]==var][\"T_VAF\"])>0.05:\n",
    "           var2use.append(var)\n",
    "ID6_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]][\n",
    "    ID6_res[\"VC\"].isin(var2use)\n",
    "].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e44bd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2use = []\n",
    "for var in set(ID9_res[\"VC\"]):\n",
    "    if max(ID9_res[ID9_res[\"VC\"]==var][\"T_VAF\"])>0.05:\n",
    "           var2use.append(var)\n",
    "ID9_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]][\n",
    "    ID9_res[\"VC\"].isin(var2use)\n",
    "].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "379069a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2use = []\n",
    "for var in set(ID10_res[\"VC\"]):\n",
    "    if max(ID10_res[ID10_res[\"VC\"]==var][\"T_VAF\"])>0.05:\n",
    "           var2use.append(var)\n",
    "ID10_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]][\n",
    "    ID10_res[\"VC\"].isin(var2use)\n",
    "].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18b9a429",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID6_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "669efcf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID9_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb52e1b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID10_res[[\"VC\", \"multiple\", \"T_VAF\", \"T_ALT\", \"sampleID_y\", \"genes\"]].sort_values(by=[\"multiple\", \"VC\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3887b6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID4_res.to_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/Results/ID4_res.csv\", index=False, header=True)\n",
    "ID5_res.to_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/Results/ID5_res.csv\", index=False, header=True)\n",
    "ID6_res.to_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/Results/ID6_res.csv\", index=False, header=True)\n",
    "ID9_res.to_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/Results/ID9_res.csv\", index=False, header=True)\n",
    "ID10_res.to_csv(\"/home/bioit/ldschaet/Desktop/FASTQ_tNGS_LA/Results/ID10_res.csv\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7226880c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID4_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ef135a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID5_res[(ID5_res[\"type\"]==\"snp\")][\"sampleID_y\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "359fb0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID6_res[(ID6_res[\"type\"]==\"snp\")][\"sampleID_y\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "380aaeb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID9_res[(ID9_res[\"type\"]==\"snp\")][\"sampleID_y\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5940bb16",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID10_res[(ID10_res[\"type\"]==\"snp\")&((ID10_res[\"T_ALT\"]>5))][\"sampleID_y\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0cf3a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID9_res[(ID9_res[\"type\"]==\"snp\")&(ID9_res[\"T_ALT\"]>5)][\"sampleID_y\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23821e1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID9_res[(ID9_res[\"type\"]==\"snp\")&(ID9_res[\"T_ALT\"]>5)][\"sampleID_y\"].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sequoia",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
